pkg/kubectl/cmd/apply.go

```diff
@@ -0,0 +1,161 @@
+		Short:   "Apply a configuration to a resource by filename or stdin",
+		Long:    apply_long,
+		Example: apply_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(validateArgs(cmd, args))
+			cmdutil.CheckErr(cmdutil.ValidateOutputArgs(cmd))
+			cmdutil.CheckErr(RunApply(f, cmd, out, options))
+		},
+	}
+
+	usage := "Filename, directory, or URL to file that contains the configuration to apply"
+	kubectl.AddJsonFilenameFlag(cmd, &options.Filenames, usage)
+	cmd.MarkFlagRequired("filename")
+	cmdutil.AddValidateFlags(cmd)
+	cmdutil.AddOutputFlagsForMutation(cmd)
+	return cmd
+}
+
+func validateArgs(cmd *cobra.Command, args []string) error {
+	if len(args) != 0 {
+		return cmdutil.UsageError(cmd, "Unexpected args: %v", args)
+	}
+
+	return nil
+}
+
+func RunApply(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *ApplyOptions) error {
+	shortOutput := cmdutil.GetFlagString(cmd, "output") == "name"
+	schema, err := f.Validator(cmdutil.GetFlagBool(cmd, "validate"), cmdutil.GetFlagString(cmd, "schema-cache-dir"))
+	if err != nil {
+		return err
+	}
+
+	cmdNamespace, enforceNamespace, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+
+	mapper, typer := f.Object()
+	r := resource.NewBuilder(mapper, typer, f.ClientMapperForCommand()).
+		Schema(schema).
+		ContinueOnError().
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		FilenameParam(enforceNamespace, options.Filenames...).
+		Flatten().
+		Latest().
+		Do()
+	err = r.Err()
+	if err != nil {
+		return err
+	}
+
+	count := 0
+	err = r.Visit(func(info *resource.Info, err error) error {
+		// In this method, info.Object contains the object retrieved from the server
+		// and info.VersionedObject contains the object decoded from the input source.
+		if err != nil {
+			return err
+		}
+
+		// Serialize the current configuration of the object from the server.
+		current, err := info.Mapping.Codec.Encode(info.Object)
+		if err != nil {
+			return cmdutil.AddSourceToErr(fmt.Sprintf("serializing current configuration from:\n%v\nfor:", info), info.Source, err)
+		}
+
+		// Retrieve the original configuration of the object from the annotation.
+		original, err := kubectl.GetOriginalConfiguration(info)
+		if err != nil {
+			return cmdutil.AddSourceToErr(fmt.Sprintf("retrieving original configuration from:\n%v\nfor:", info), info.Source, err)
+		}
+
+		// Get the modified configuration of the object. Embed the result
+		// as an annotation in the modified configuration, so that it will appear
+		// in the patch sent to the server.
+		modified, err := kubectl.GetModifiedConfiguration(info, true)
+		if err != nil {
+			return cmdutil.AddSourceToErr(fmt.Sprintf("retrieving modified configuration from:\n%v\nfor:", info), info.Source, err)
+		}
+
+		// Compute a three way strategic merge patch to send to server.
+		patch, err := strategicpatch.CreateThreeWayMergePatch(original, modified, current, info.VersionedObject, false)
```

Comment:
Another thing : did you intend to do the 3way merge patch with the "original" annotation still in the object? I would think the correct procedure would be to do the threeway merge patch without the `latest-applied-configuration`, then insert it into the patch afterwards. i.e.
1. extract `original` from `current`'s annotation
2. delete annotation from `current`
3. threewaymergepatch original, modified, current
4. add annotation = modified to the patch

Could also be you worked thru this case and merging the annotation is correct. If so, it deserves a comment explaining why.